[ews-build.webkit.org] Upload full build logs to S3

JonWBedard opened the pull request:
#### 4f987fb3ae60485e3e0f44f82984caa301210c20
<pre>
[ews-build.webkit.org] Upload full build logs to S3
<a href="https://bugs.webkit.org/show_bug.cgi?id=268766">https://bugs.webkit.org/show_bug.cgi?id=268766</a>
<a href="https://rdar.apple.com/122328282">rdar://122328282</a>

Reviewed by Aakash Jain.

To prevent buildbot from attempting to consume each line of a build log one-by-one,
we should filter logs through filter-build-webkit and then upload those logs to S3.

* Tools/CISupport/ews-build/steps.py:
(CompileWebKit): Separate build and filter commands.
(CompileWebKit.start): Concatenate build and filter commands.
(CompileWebKit.follow_up_steps): Upload build logs to S3 for Apple&apos;s ports.
(CompileWebKit.evaluateCommand): In production and staging instances for Apple&apos;s ports,
upload logs to S3.
(CompileWebKitWithoutChange.evaluateCommand): Ditto.
(CompileJSC): Separate build and filter commands.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/274168@main">https://commits.webkit.org/274168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e9c04a2355fa18f6a10715015ad518065579b9af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4955 "Built successfully and passed tests") | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3130,7 +3130,9 @@ class CompileWebKit(shell.Compile, AddToLogMixin):
     warningPattern = '.*arning: .*'
     haltOnFailure = False
-    command = ['perl', 'Tools/Scripts/build-webkit', WithProperties('--%(configuration)s')]
+    build_command = ['perl', 'Tools/Scripts/build-webkit']
```
any specific reason to remove "WithProperties('--%(configuration)s')" and add it later inside start()?

JonWBedard replied:
This command gets joined instead passed as a string to the build command, so we have to resolve `WithProperties` before we pass the command to our parent class.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3130,7 +3130,9 @@ class CompileWebKit(shell.Compile, AddToLogMixin):
-    command = ['perl', 'Tools/Scripts/build-webkit', WithProperties('--%(configuration)s')]
+    build_command = ['perl', 'Tools/Scripts/build-webkit']
+    filter_command = ['perl', 'Tools/Scripts/filter-build-webkit', '-logfile', 'WebKitBuild/build-log.txt']
+    command = ['/bin/sh', '-c']
```
Do we actually need to set this command here? it seems weird/convoluted to set half of the command here (which gives no indication of what it will do), and then complete it later. Might as well just set it once later-on.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3469,7 +3488,7 @@ def send_email_for_preexisting_build_failure(self):
     name = 'compile-jsc'
     descriptionDone = ['Compiled JSC']
-    command = ['perl', 'Tools/Scripts/build-jsc', WithProperties('--%(configuration)s')]
+    build_command = ['perl', 'Tools/Scripts/build-jsc']
```
Ditto. Do we need to remove WithProperties('--%(configuration)s')?

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4938,6 +4957,27 @@ def getResultSummary(self):
+class ZipFile(shell.ShellCommandNewStyle):
+    name = 'zip-file'
+    descriptionDone = name
+    haltOnFailure = False
```
Just to mention, haltOnFailure is false here, but failure of this step will cause subsequent UploadFileToS3() to fail which has haltOnFailure=True.

JonWBedard replied:
Yes, I would like to address this in a later change. I'm aware that failing to upload a log will fail the build. We need to make `UploadFileToS3` `haltOnFailure` only when uploading a build archive, but not when uploading a log.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3151,33 +3153,39 @@ def start(self):
+
+        if platform in self.APPLE_PLATFORMS:
+            self.setCommand(self.shell_command(f"{' '.join(build_command)} | {' '.join(self.filter_command)}"))
+        else:
```
Might be a good idea to add a comment here mentioning the reason for different behaviour on apple and non-apple platforms.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3123,15 +3123,17 @@ def parseOutputLine(self, line):
+    build_command = ['perl', 'Tools/Scripts/build-webkit', WithProperties('--%(configuration)s')]
+    filter_command = ['perl', 'Tools/Scripts/filter-build-webkit', '-logfile', 'WebKitBuild/build-log.txt']
     VALID_ADDITIONAL_ARGUMENTS_LIST = []  # If additionalArguments is added to config.json for CompileWebKit step, it should be added here as well.
+    APPLE_PLATFORMS = ('mac', 'ios', 'tvos', 'watchos')
```
It might be better to make a list of Non-Apple platforms since that list will be smaller and less likely to change.

Also, this excludes queues with platform as '*', which is fine currently, but in future it might change.

JonWBedard committed:
[ews-build.webkit.org] Upload full build logs to S3
https://bugs.webkit.org/show_bug.cgi?id=268766
rdar://122328282

Reviewed by Aakash Jain.

To prevent buildbot from attempting to consume each line of a build log one-by-one,
we should filter logs through filter-build-webkit and then upload those logs to S3.

* Tools/CISupport/ews-build/steps.py:
(CompileWebKit): Separate build and filter commands.
(CompileWebKit.start): Concatenate build and filter commands.
(CompileWebKit.follow_up_steps): Upload build logs to S3 for Apple's ports.
(CompileWebKit.evaluateCommand): In production and staging instances for Apple's ports,
upload logs to S3.
(CompileWebKitWithoutChange.evaluateCommand): Ditto.
(CompileJSC): Separate build and filter commands.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/274168@main
